### PR TITLE
Add dynamic minimum fan speed support

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
**- What I did**

1. Add a fan fault policy to check fan fault. If any fan fault, set fan speed to 100% and disable kernel algorithm
2. Add dynamic minimum fan speed support, minimum fan speed will be determined by air direction, highest temperature and sensor trust state.

**- How I did it**

1. Add a condition to check fan fault and change the default policy file
2. Add a condition to check if dynamic minimum fan speed need change and an action to change minimum fan speed
3. Add a policy to update dynamic minimum fan speed every 60 seconds, this policy is private that user cannot configure it

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by keboliu** [comment]
can we split this PR to two, each topic has a separated PR?

File: platform/mellanox/mlnx-platform-api/sonic_platform/device_data.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                "p2c_trust":   {"-127:35":12, "36:120":13},
+                "p2c_untrust": {"-127:0":12, "1:10":13 , "11:15":14, "16:20":15, "21:35":16, "36:120":17},
+                "c2p_trust":   {"-127:30":12, "31:40":13 , "41:120":14},
+                "c2p_untrust":   {"-127:20":12, "21:30":13 , "31:35":14, "36:40":15, "41:120":16},
+                "unk_trust":   {"-127:30":12, "31:40":13 , "41:120":14},
+                "unk_untrust": {"-127:0":12, "1:10":13 , "11:15":14, "16:20":15, "21:35":16, "36:120":17},
+            }
+        }
+    },
+}
```
Comment (keboliu) [inline platform/mellanox/mlnx-platform-api/sonic_platform/device_data.py:134]: I know we don't have the table for 4700 yet, what's the behavior on 4700 w/o this table?

Comment (Junchao-Mellanox) [reply platform/mellanox/mlnx-platform-api/sonic_platform/device_data.py:134]: minimum FAN speed of 4700 is 60%

File: platform/mellanox/mlnx-platform-api/sonic_platform/fan.py [hunk 2]
```diff
@@ -24,6 +24,11 @@
 LED_PATH = "/var/run/hw-management/led/"
 # fan_dir isn't supported on Spectrum 1. It is supported on Spectrum 2 and later switches
 FAN_DIR = "/var/run/hw-management/system/fan_dir"
+COOLING_STATE_PATH = "/var/run/hw-management/thermal/cooling_cur_state"
+
+# SKUs with unplugable FANs:
+# 1. don't have fanX_status and should be treated as always present
+hwsku_dict_with_unplugable_fan = ['ACS-MSN2010', 'ACS-MSN2100']
 
 # SKUs with unplugable FANs:
```
Comment (keboliu) [inline platform/mellanox/mlnx-platform-api/sonic_platform/fan.py:33]: remove redundant comments

File: platform/mellanox/mlnx-platform-api/sonic_platform/fan.py [hunk 3]
```diff
@@ -352,3 +363,36 @@ def get_speed_tolerance(self):
...
+
+        if level < 1 or level > 10:
+            raise RuntimeError("Failed to set cooling level, level value must be in range [1, 10], got {}".format(level))
+
+        try:
+            # reset FAN driver and change cooling state
+            with open(COOLING_STATE_PATH, 'w') as cooling_state:
+                cooling_state.write(str(level + 10))
+
+            # make cooling state diplay correct value
```
Comment (keboliu) [inline platform/mellanox/mlnx-platform-api/sonic_platform/fan.py:384]: "display"

File: platform/mellanox/mlnx-platform-api/sonic_platform/thermal_actions.py [hunk 4]
```diff
@@ -95,14 +111,47 @@ def execute(self, thermal_info_dict):
         :param thermal_info_dict: A dictionary stores all thermal information.
         :return:
         """
+        from .thermal_infos import FanInfo
+        from .thermal import Thermal
+        from .fan import Fan
+        Thermal.set_thermal_algorithm_status(self.status, False)
+        if self.status:
+            # Check thermal zone temperature, if all thermal zone temperature
+            # return to normal and FAN speed is still 100%, set it to 60% to
```
Comment (keboliu) [inline platform/mellanox/mlnx-platform-api/sonic_platform/thermal_actions.py:120]: "return to normal" --> "back to normal" ?

File: platform/mellanox/mlnx-platform-api/sonic_platform/device_data.py (hunk 1 above)
Comment (keboliu) [reply platform/mellanox/mlnx-platform-api/sonic_platform/device_data.py:134]: we should follow up when can we have the table for 4700.

Comment (Junchao-Mellanox) [reply platform/mellanox/mlnx-platform-api/sonic_platform/device_data.py:134]: sure

**Comment by Junchao-Mellanox** [comment]
Community PR https://github.com/Azure/sonic-buildimage/pull/4403

**Pull request closed without merging**